File: dart/reverse_linked_list.dart
class ListNode {
  int? val;
  ListNode? next;

  ListNode([this.val, this.next]);
}

class Solution {
  // Method to reverse a linked list.
  ListNode? reverseList(ListNode? head) {
    // Declare two pointers, prev and curr.
    ListNode? prev, curr;
    prev =
        null; // Initialize prev to null as there's no node before the head initially.
    curr = head; // Initialize curr to the head of the linked list.

    // Iterate through the linked list.
    while (curr != null) {
      // Save the reference to the next node.
      ListNode? nxt = curr.next;

      // Reverse the link direction.
      curr.next = prev;

      // Move prev and curr pointers to the next nodes.
      prev = curr;
      curr = nxt;
    }

    // The new head of the reversed linked list is the last node.
    return prev;
  }
}

File: python/reverse_linked_list.py
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Reverses a singly linked list in place.

        Args:
            head: The head of the linked list to be reversed.

        Returns:
            The head of the reversed linked list.
        """

        prev = None  # Pointer to the previous node (starts as None)
        curr = head  # Pointer to the current node (starts at the head)

        while curr:  # Iterate through the list until reaching the end
            nxt = curr.next  # Store the next node before reversing links
            curr.next = prev  # Reverse the link of the current node
            prev = curr  # Move prev to the current node for the next iteration
            curr = nxt  # Move curr to the next node

        return prev  # The new head of the reversed list is the last node in the original list
